### PR TITLE
port(turborepo): Paths Crate

### PR DESCRIPTION
**Pull request by NicholasLYang**
### Description

Decided to adapt buck2's path module as our `turborepo-paths` crate. This PR moves it over and adjusts some naming/style choices.

### Testing Instructions

The module came with some pretty thorough tests and doctests. You can run with `cargo test -p turborepo-paths`.

- [x] Auto label

### CODE REVIEW

File: crates/turborepo-paths/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+name = "turborepo-paths"
+version = "0.1.0"
+edition = "2021"
+license = "MIT"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+allocative = "0.2.0"
+anyhow = "1.0.69"
```
Comment (ForsakenHarmony) [inline crates/turborepo-paths/Cargo.toml:11]: reuse the common dependencies from the workspace :p

**Review by mehulkar — COMMENTED** [review]
> Decided to adapt buck2's path module as our turborepo-paths crate. This PR moves it over and adjusts some naming/style choices.

Is this an external library that's vendored and adjusted? Is it possible to add as a dependency and extend/wrap instead?

File: crates/turborepo-paths/src/absolute_normalized_path.rs [hunk 2]
```diff
@@ -0,0 +1,982 @@
...
+    /// # use turborepo_paths::absolute_normalized_path::AbsoluteNormalizedPath;
+    ///
+    /// assert!(AbsoluteNormalizedPath::new("foo/bar").is_err());
+    /// if cfg!(windows) {
+    ///    assert!(AbsoluteNormalizedPath::new("C:\\foo\\bar").is_ok());
+    /// } else {
+    ///    assert!(AbsoluteNormalizedPath::new("/foo/bar").is_ok());
+    /// }
+    /// # anyhow::Ok(())
+    /// ```
```
[the remark this answers is not in the document]
Comment (NicholasLYang) [reply crates/turborepo-paths/src/absolute_normalized_path.rs:112]: These are doc tests. Basically they're inline tests that also function as examples for people reading documentation. They get built and run with `cargo test`

File: crates/turborepo-paths/LICENSE [hunk 3]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) Meta Platforms, Inc. and affiliates.
```
[the remark this answers is not in the document]
Comment (NicholasLYang) [reply crates/turborepo-paths/LICENSE:3]: That's good question. @gsoltis any advice?

**Comment by NicholasLYang** [comment]
The library was a module inside buck2's core package, so unless we wanted to have buck2_core as a dependency (and even then not all of the paths components may be exported from the crate), we can't really wrap it. Anyways buck2_core isn't exported as a crate, so we'd have to use a git dependency which is not ideal because git dependencies pull down the entire repo which makes building quite slow.

File: crates/turborepo-paths/LICENSE (hunk 3 above)
[the remark this answers is not in the document]
Comment (arlyon) [reply crates/turborepo-paths/LICENSE:3]: I think we need to keep the copyright in the files and retain the crate license field. We can probably remove the license file.

**Review by arlyon — APPROVED** [review]
Satisfyingly few changes needed :) It may be an idea to list any major changes made to the module in lib.rs (and maybe change the 'for buck2' part)

**Pull request merged**